aten/src/ATen/templates/TensorBody.h

```diff
@@ -394,13 +447,178 @@ class CAFFE2_API Tensor {
     return impl_->requires_grad();
   }
 
+  // TODO: There is something bad about this: const-correctness on Tensor
+  // is not supposed to matter, but here we are looking at the constness
+  // of Tensor to decide whether or not to return a mutable reference...
   Tensor& grad() {
     return impl_->grad();
   }
   const Tensor& grad() const {
     return impl_->grad();
   }
 
+  /// NOTE: This is similar to the legacy `.data()` function on `Variable`, and is intended
+  /// to be used from functions that need to access the `Variable`'s equivalent `Tensor`
+  /// (i.e. `Tensor` that shares the same storage and tensor metadata with the `Variable`).
+  ///
+  /// One notable difference with the legacy `.data()` function is that changes to the
+  /// returned `Tensor`'s tensor metadata (e.g. sizes / strides / storage / storage_offset)
+  /// will not update the original `Variable`, due to the fact that this function
+  /// shallow-copies the `Variable`'s underlying TensorImpl.
+  Tensor tensor_data() const noexcept {
```

Comment:
@yf225 So, what's the situation here? Is the idea that once `data()` goes through long enough deprecation cycle, we rename `tensor_data()` to `data()`, replacing the old implementation?

Reply:
To put it differently, do we want `tensor_data()` and `variable_data()` in our API? At the very least, their semantics is very confusing to me.

Reply:
From what I understand of these api, `tensor_data() == detach()` and `variable_data() == .data` in python.

Reply:
The semantics is that `(C++) variable_data() == (Python) .data`, and `(C++) detach() == (Python) detach()`.

`tensor_data()` tries to mimic the old semantics of `.data()` in the "Variable wrapping Tensor" days (i.e. it returns a shallow-copy of the Variable with a null AutogradMeta), because some call sites of `tensor_data()` rely on this semantics. We actually want to expunge all use cases of `tensor_data()` in our codebase and changing them to `variable_data()`. I worked on a patch (https://github.com/pytorch/pytorch/pull/22172) before but we haven't got to a point where it's ready to merge :(

Reply:
Once we expunge all `tensor_data()` call sites and changing them to `variable_data()`, we can rename `variable_data()` to `data()`, which would be consistent with Python API.